aws-s3-transfer-manager/src/operation/upload_objects/handle.rs

```diff
@@ -3,17 +3,43 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-use super::{UploadObjectsError, UploadObjectsOutput};
+use std::sync::atomic::Ordering;
+
+use super::{UploadObjectsContext, UploadObjectsOutput};
+use tokio::task;
 
 /// Handle for `UploadObjects` operation
 #[derive(Debug)]
 #[non_exhaustive]
-pub struct UploadObjectsHandle {}
+pub struct UploadObjectsHandle {
+    /// All child tasks spawned for this download
+    pub(crate) tasks: task::JoinSet<Result<(), crate::error::Error>>,
+    /// The context used to drive an upload to completion
+    pub(crate) ctx: UploadObjectsContext,
+}
 
 impl UploadObjectsHandle {
     /// Consume the handle and wait for the upload to complete
     #[tracing::instrument(skip_all, level = "debug", name = "join-upload-objects-")]
-    pub async fn join(self) -> Result<UploadObjectsOutput, UploadObjectsError> {
-        unimplemented!()
+    pub async fn join(mut self) -> Result<UploadObjectsOutput, crate::error::Error> {
+        while let Some(join_result) = self.tasks.join_next().await {
+            join_result??;
+        }
+
+        let failed_uploads = self.ctx.state.failed_uploads.lock().unwrap().take();
+        let successful_uploads = self.ctx.state.successful_uploads.load(Ordering::SeqCst);
+        let total_bytes_transferred = self
+            .ctx
+            .state
+            .total_bytes_transferred
+            .load(Ordering::SeqCst);
+
+        let output = UploadObjectsOutput::builder()
+            .objects_uploaded(successful_uploads)
+            .set_failed_transfers(failed_uploads)
+            .total_bytes_transferred(total_bytes_transferred)
+            .build();
```

Comment:
This seems like a use case for
```impl From<UploadObjectsState> for UploadObjectOutput```. Any reason not to use that here?

Reply:
Good suggestion! Since `UploadObjectsState` is behind `Arc` what we can obtain is `&UploadObjectsState`. Added `impl From` for both upload objects and download objects in e558339.